In [2]:
import numpy as np
import pandas as pd

# Display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:.2f}".format)

In [3]:
# loading data set
df = pd.read_csv("../data/BankChurners.csv")
df.sample(5)



CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
3686  708211383  Existing Customer            35      F                3   
370   719091333  Existing Customer            52      M                0   
777   828298908  Existing Customer            54      M                4   
3775  807596358  Existing Customer            49      M                2   
3231  771081558  Existing Customer            55      M                2   

     Education_Level Marital_Status Income_Category Card_Category  \
3686        Graduate         Single  Less than $40K          Blue   
370       Uneducated        Married     $60K - $80K          Blue   
777         Graduate        Married         $120K +          Blue   
3775      Uneducated        Married    $80K - $120K          Blue   
3231     High School       Divorced         $120K +          Blue   

      Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
3686              27                         3                       3   
370               36                         3                       3   
777               50                         4                       2   
3775              44                         6                       2   
3231              42                         3                       2   

      Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
3686                      3       2664.00                    0   
370                       2      22720.00                  793   
777                       0       2872.00                 2035   
3775                      2      13884.00                 1204   
3231                      2      34516.00                 1190   

      Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
3686          2664.00                  0.73             3906              74   
370          21927.00                  0.69             1102              27   
777            837.00                  0.61             1770              47   
3775         12680.00                  0.57             4650              70   
3231         33326.00                  0.83             2862              78   

      Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
3686                 0.57                   0.00   
370                  0.29                   0.04   
777                  0.74                   0.71   
3775                 0.56                   0.09   
3231                 0.73                   0.03   

      Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
3686                                               0.00                                                                                    
370                                                0.00                                                                                    
777                                                0.00                                                                                    
3775                                               0.00                                                                                    
3231                                               0.00                                                                                    

      Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
3686                                               1.00                                                                                   
370                                                1.00                                                                                   
777                                                1.00                                                                                   
3775                                               1.00                                              

In [4]:
# shape and info of our df 
print(df.shape)
print(df.info())

(10127, 23)
<class 'pandas.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  str    
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                          

- The dataset contains 10127 rows and 23 columns.
- Each row represents a unique customer.
- Numerical columns: int64, float64
- Categorical columns: object
No immediate type inconsistencies detected.


In [ ]:
# Mising value analysis
df.isnull().sum()


CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

No missing values detected in the dataset.
Data is clean and ready for analysis.


In [ ]:
# Duplicate Records Check
df.duplicated().sum()

np.int64(0)

No duplicate records found.
Each customer record appears unique.

In [ ]:
# Separate Numerical & Categorical Columns
numerical_cols = df.select_dtypes(include=['int64', 'float64'])
categorical_cols = df.select_dtypes(include=['object'])

C:\Users\aazib\AppData\Local\Temp\ipykernel_3736\3977160696.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object'])


In [11]:
# Numerical Summary
numerical_summary = numerical_cols.describe().T
numerical_summary.round(2)

count         mean  \
CLIENTNUM                                          10127.00 739177606.33   
Customer_Age                                       10127.00        46.33   
Dependent_count                                    10127.00         2.35   
Months_on_book                                     10127.00        35.93   
Total_Relationship_Count                           10127.00         3.81   
Months_Inactive_12_mon                             10127.00         2.34   
Contacts_Count_12_mon                              10127.00         2.46   
Credit_Limit                                       10127.00      8631.95   
Total_Revolving_Bal                                10127.00      1162.81   
Avg_Open_To_Buy                                    10127.00      7469.14   
Total_Amt_Chng_Q4_Q1                               10127.00         0.76   
Total_Trans_Amt                                    10127.00      4404.09   
Total_Trans_Ct                                     10127.00        64.86   
Total_Ct_Chng_Q4_Q1                                10127.00         0.71   
Avg_Utilization_Ratio                              10127.00         0.27   
Naive_Bayes_Classifier_Attrition_Flag_Card_Cate... 10127.00         0.16   
Naive_Bayes_Classifier_Attrition_Flag_Card_Cate... 10127.00         0.84   

                                                           std          min  \
CLIENTNUM                                          36903783.45 708082083.00   
Customer_Age                                              8.02        26.00   
Dependent_count                                           1.30         0.00   
Months_on_book                                            7.99        13.00   
Total_Relationship_Count                                  1.55         1.00   
Months_Inactive_12_mon                                    1.01         0.00   
Contacts_Count_12_mon                                     1.11         0.00   
Credit_Limit                                           9088.78      1438.30   
Total_Revolving_Bal                                     814.99         0.00   
Avg_Open_To_Buy                                        9090.69         3.00   
Total_Amt_Chng_Q4_Q1                                      0.22         0.00   
Total_Trans_Amt                                        3397.13       510.00   
Total_Trans_Ct                                           23.47        10.00   
Total_Ct_Chng_Q4_Q1                                       0.24         0.00   
Avg_Utilization_Ratio                                     0.28         0.00   
Naive_Bayes_Classifier_Attrition_Flag_Card_Cate...        0.37         0.00   
Naive_Bayes_Classifier_Attrition_Flag_Card_Cate...        0.37         0.00   

                                                            25%          50%  \
CLIENTNUM                                          713036770.50 717926358.00   
Customer_Age                                              41.00        46.00   
Dependent_count                                            1.00         2.00   
Months_on_book                                            31.00        36.00   
Total_Relationship_Count                                   3.00         4.00   
Months_Inactive_12_mon                                     2.00         2.00   
Contacts_Count_12_mon                                      2.00         2.00   
Credit_Limit                                            2555.00      4549.00   
Total_Revolving_Bal                                      359.00      1276.00   
Avg_Open_To_Buy                                         1324.50      3474.00   
Total_Amt_Chng_Q4_Q1                                       0.63         0.74   
Total_Trans_Amt                                         2155.50      3899.00   
Total_Trans_Ct                                            45.00        67.00   
Total_Ct_Chng_Q4_Q1                                        0.58         0.70   
Avg_Utilization_Ratio                                      0.0

In [12]:
# Catagorical Summary
categorical_summary = categorical_cols.describe().T
categorical_summary


count unique                top  freq
Attrition_Flag   10127      2  Existing Customer  8500
Gender           10127      2                  F  5358
Education_Level  10127      7           Graduate  3128
Marital_Status   10127      4            Married  4687
Income_Category  10127      6     Less than $40K  3561
Card_Category    10127      4               Blue  9436

In [14]:
# Target Variable Distribution
df['Attrition_Flag'].value_counts()
df["Attrition_Flag"].value_counts(normalize=True) * 100

Attrition_Flag
Existing Customer   83.93
Attrited Customer   16.07
Name: proportion, dtype: float64

## Variable Classification

Identifier:
- CLIENTNUM

Target:
- Attrition_Flag

Demographic:
- Customer_Age
- Gender
- Dependent_count
- Education_Level
- Marital_Status
- Income_Category

Relationship:
- Months_on_book
- Total_Relationship_Count

Behavioral:
- Months_Inactive_12_mon
- Contacts_Count_12_mon
- Total_Trans_Ct
- Total_Ct_Chng_Q4_Q1
- Total_Amt_Chng_Q4_Q1

Financial Exposure:
- Credit_Limit
- Total_Revolving_Bal
- Avg_Open_To_Buy
- Total_Trans_Amt
- Avg_Utilization_Ratio
